### DISCUSSION TITLE
I want to update to hand refresh, but the effect is one-time

### DISCUSSION THREAD

**Discussion by shuangmuyuan** [argoproj/argo-cd#7737] [Q&A]
argocd version：
Argo CD
v2.1.7+a408e29
Helm
v3.6.0+g7f2df64
kubectl
v0.21.0
Problem Description：
`kubectl patch app/atest-api -n argocd --type merge -p='{"metadata": {"annotations":{"argocd.argoproj.io/refresh": "hard"}}}'`
It doesn’t seem to work when I use patch：
`kubectl get app/atest-api -n argocd -o yaml`
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  creationTimestamp: "2021-11-15T11:37:40Z"
  generation: 1390
  name: atest-api
  namespace: argocd
  resourceVersion: "23013509"
  uid: d80322cb-c8c0-4ba9-8597-44d0bcef4995
spec:
  .....
```
But I saw the log did trigger a hard refresh（argocd-application-controller's log）
```
time="2021-11-18T06:49:47Z" level=info msg="Refreshing app status (hard refresh requested), level (3)" application=atest-api
time="2021-11-18T06:49:47Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-11-18T06:49:47Z" level=info msg="Ignore '/spec/preserveUnknownFields' for CustomResourceDefinitions"
time="2021-11-18T06:49:47Z" level=info msg="Comparing app state (cluster: https://10.134.85.75:6443, namespace: nwj-test)" application=asoc-api
time="2021-11-18T06:49:47Z" level=info msg="getRepoObjs stats" application=asoc-api build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=373 unmarshal_ms=372 version_ms=0
time="2021-11-18T06:49:47Z" level=info msg="Update successful" application=atest-api
```
my expectation is that this project will trigger hard refresh afterwards, but it does not seem to be the case.
I edit the ANNOTATIONS of the ui application, and set argocd.argoproj.io/refresh to hard is also a one-time, and it is not saved in the configuration file

**Comment by pasha-codefresh** [comment]
According to comment here https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/application_annotations.go#L6
looks like it is expected behavior

**Comment by pmorch** [comment]
See also #7774, were @jessesuen writes on Dec 2, 2021:

> This was discussed in today's contributors meeting. Providing a control knob to perform hard refreshes with every reconciliation will have a performance impact which we don't want to incur. Argo CD design is to be able to reconcile an application early and often, and quickly with little to no performance consequence (in the event that nothing changed). By adding an option to always perform hard refreshes for applications, it causes the application to do a large amount of work with every reconciliation, and block the limited workers in the controller.
> 
> > My application is deployed using helm. The tag of helm will not change, but the image will change. Argocd cannot automatically check the changes of helm.
> 
> This is a practice which we highly discourage. GitOps expects semantic versions (as well as git commit SHAs) to be immutable and produce the same output of manifests. By having the contents of the manifests change, with no modification to the application spec or git repo, it breaks GitOps practices and makes rollbacks essentially meaningless.

**Reply by jbartyze-rh** [reply]
Do you guys consider vaultplugin interaction in this discussion? 

What I observed is that ArgoCD + vault plugin may based on stale cache introduce old stale secrets to kubernetes/OpenShift cluster. 

So for example OpenShift Machineset object will inject with from vault clusterid during manifest generation. If repo cache was generated before cluster was rebuilded and new entry was introduced on vault side it will have new clusterid entry in vault and old in cache. This will cause argocd to sync old generated manifest and Machineset with incorrect clusterid. 

Only way to ensure that this does not happen is to trigger hard refresh or change repo server cache values to really low time(1 minute?). I wonder how this use case would best solved.

I fully agree for other use cases it is desirable to sync from cache to preserve upstream Git Repo performance, but we need ensure that our interaction with Vault Plugin will always refresh secret from vault.
